### ISSUE TITLE
undefined references to all the methods in libtorch

### ISSUE THREAD

**Issue by lucasjinreal** [fabvio/LD_CNN_classic#1]
```
CMakeFiles/lane_cnn.dir/main.cpp.o: In function `at::detail::infer_type(at::Tensor const&)':
main.cpp:(.text+0xfa8): undefined reference to `at::getType(at::Tensor const&)'
CMakeFiles/lane_cnn.dir/main.cpp.o: In function `at::empty(c10::ArrayRef<long>, c10::TensorOptions const&)':
main.cpp:(.text+0x105d): undefined reference to `at::getType(c10::TensorOptions)'
CMakeFiles/lane_cnn.dir/main.cpp.o: In function `main':
main.cpp:(.text+0x2142): undefined reference to `torch::jit::load(std::string const&, c10::optional<c10::Device>, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > >&)'
CMakeFiles/lane_cnn.dir/main.cpp.o: In function `at::LegacyTypeDispatch::getTypeRaw(c10::Backend, c10::ScalarType, bool)':
main.cpp:(.text._ZN2at18LegacyTypeDispatch10getTypeRawEN3c107BackendENS1_10ScalarTypeEb[_ZN2at18LegacyTypeDispatch10getTypeRawEN3c107BackendENS1_10ScalarTypeEb]+0x37): undefined reference to `at::detail::getVariableHooks()'
CMakeFiles/lane_cnn.dir/main.cpp.o: In function `at::legacyTensorType(c10::TensorImpl const&)':
main.cpp:(.text._ZN2at16legacyTensorTypeERKN3c1010TensorImplE[_ZN2at16legacyTensorTypeERKN3c1010TensorImplE]+0x4e): undefined reference to `at::globalLegacyTypeDispatch()'
CMakeFiles/lane_cnn.dir/main.cpp.o: In function `at::from_blob(void*, c10::ArrayRef<long>, c10::ArrayRef<long>, std::function<void (void*)> const&, c10::TensorOptions const&)':
main.cpp:(.text._ZN2at9from_blobEPvN3c108ArrayRefIlEES3_RKSt8functionIFvS0_EERKNS1_13TensorOptionsE[_ZN2at9from_blobEPvN3c108ArrayRefIlEES3_RKSt8functionIFvS0_EERKNS1_13TensorOptionsE]+0x72): undefined reference to `at::getType(c10::TensorOptions)'
main.cpp:(.text._ZN2at9from_blobEPvN3c108ArrayRefIlEES3_RKSt8functionIFvS0_EERKNS1_13TensorOptionsE[_ZN2at9from_blobEPvN3c108ArrayRefIlEES3_RKSt8functionIFvS0_EERKNS1_13TensorOptionsE]+0x22e): undefined reference to `at::detail::computeStorageSize(c10::ArrayRef<long>, c10::ArrayRef<long>)'
CMakeFiles/lane_cnn.dir/main.cpp.o: In function `at::from_blob(void*, c10::ArrayRef<long>, std::function<void (void*)> const&, c10::TensorOptions const&)':
main.cpp:(.text._ZN2at9from_blobEPvN3c108ArrayRefIlEERKSt8functionIFvS0_EERKNS1_13TensorOptionsE[_ZN2at9from_blobEPvN3c108ArrayRefIlEERKSt8functionIFvS0_EERKNS1_13TensorOptionsE]+0x43): undefined reference to `at::detail::defaultStrides(c10::ArrayRef<long>)'
CMakeFiles/lane_cnn.dir/main.cpp.o: In function `c10::FunctionSchema::checkArg(c10::IValue const&, c10::Argument const&, c10::optional<unsigned long>) const':

```

**Comment by fabvio** [comment]
Have you tried generating the makefile with cmake -DCMAKE_PREFIX_PATH=path/to/libtorch .. ?

**Comment by lucasjinreal** [comment]
@fabvio  tried, I copied codes into a fresh workspace and it can built without those headers. BTW, where does the model comes from? It outputs nothing on my lane image?

**Comment by fabvio** [comment]
The model is a torch::jit traced version of ERFNet trained on the TuSimple dataset for lane detection. If you have no output it can be caused by two factors:

1. The input you are passing to the network is not normalized correctly
2. The image is too different from the ones of the training set, so the domain shift is too wide to be managed by the CNN.

Could you please upload a sample image? Also, if you are interested in a PyTorch version of the network, I will release in a few days the code for the paper this network has been trained for, so follow me to be updated.

**Comment by lucasjinreal** [comment]
@fabvio  thanks so much! I tested on this image, it show  out some lines (location does not correct):

![town01501](https://user-images.githubusercontent.com/21303438/58603315-e17f6680-82c2-11e9-8078-900b517ca1a0.jpg)


Does it because it's too small? Hoping for your pytorch training script release. C++ version I can run it now. But this image prediction location is not right (above the lane a lot). What's your prediction result?

**Comment by fabvio** [comment]
I cannot run it now as at the moment I have no access to the cluster and to my files. I suppose that the domain shift between the images of the training set and you image is too broad. Take into account that the CNN here has been trained on the TuSimple lane detection dataset, that only includes highway scenarios. Also, the instance segmentation loss makes it somehow more fragile to outliers.

I suggest you to train a CNN on the CuLane dataset for lane detection, or, if you are looking to alternatives, with [this](https://github.com/fabvio/ld-lsi) CNN you should be able to obtain a correct output on freespace on multiple lanes, maybe you can infer the lane information from that one with post processing.
